wandb/sdk/launch/_project_spec.py

```diff
@@ -487,30 +487,28 @@ def get_env_vars_dict(self, api: Api, max_env_length: int) -> Dict[str, str]:
         return env_vars
 
     def parse_existing_requirements(self) -> str:
-        import pkg_resources
+        from packaging.requirements import InvalidRequirement, Requirement
 
         requirements_line = ""
         assert self.project_dir is not None
         base_requirements = os.path.join(self.project_dir, "requirements.txt")
         if os.path.exists(base_requirements):
             include_only = set()
-            with open(base_requirements) as f:
-                iter = pkg_resources.parse_requirements(f)
-                while True:
+            with open(base_requirements) as f2:
+                for line in f2:
+                    if line.strip() == "":
+                        continue
+
                     try:
-                        pkg = next(iter)
-                        if hasattr(pkg, "name"):
-                            name = pkg.name.lower()
-                        else:
-                            name = str(pkg)
+                        req = Requirement(line)
+                        name = req.name.lower()
                         include_only.add(shlex_quote(name))
-                    except StopIteration:
-                        break
-                    # Different versions of pkg_resources throw different errors
-                    # just catch them all and ignore packages we can't parse
-                    except Exception as e:
-                        _logger.warning(f"Unable to parse requirements.txt: {e}")
+                    except InvalidRequirement as e:
+                        _logger.warning(
+                            f"Unable to parse line {line} in requirements.txt: {e}"
+                        )
```

Comment:
We don't have the lint enabled, and we break this rule more often than follow it, but the recommended way to log interpolated strings is to like so:

```python
_logger.warning("Unable to parse line %s in requirements.txt", line)
```

This avoids string conversion and interpolation if WARNING-level logs are ignored.

I think you also shouldn't need to include the exception explicitly; `warning()` will include it automatically. It would be good to double check though. If you use `caplog` in the test instead of mocking, you'll be able to see the real output.

```python
except InvalidRequirement:
    _logger.warning("Unable to parse line %s in requirements.txt", line)
```

Reply:
The lint I'm referring to, with more info: https://docs.astral.sh/ruff/rules/logging-f-string/

Reply:
And if `warning` doesn't automatically read the exception info, pass `exc_info=True`